rdt/transformers/categorical.py

```diff
@@ -455,14 +477,39 @@ def _fit(self, data):
     def _transform(self, data):
         """Replace each category with its corresponding integer value.
 
+        If a category has not been seen before, a random value is assigned.
+        If no categories have been fitted (and `data` is not empty), raise a `ValueError`.
+
         Args:
             data (pandas.Series):
                 Data to transform.
 
         Returns:
-            numpy.ndarray:
+            pd.Series.
         """
-        return pd.Series(data).map(self.categories_to_values)
+        # Find the data categories that were not seen during fitting. Convert all nan's
+        # to np.nan to avoid having nan values with different ids.
+        unique_data = _select_unique_categories(pd.unique(data))
+        categories = _select_unique_categories(pd.Series(self.categories_to_values.keys()))
+        unseen_categories = unique_data - categories
+        if unseen_categories:
+            if len(self.categories_to_values) == 0:
```

Comment:
@npatki indicated that we shouldn't be worrying about fitting before transforming in this PR. In the future we may go through the models again to assert you don't transform without fitting, but for now I shouldn't make any changes about it.

In this PR though, since it hasn't been discussed what the behavior should be if we fit on an empty dataframe, I added that guarantee there.

Reply:
I wouldn't worry about that in this PR. I also am not sure that's worth raising an error for.

Reply:
+1 In the case where a user accidentally tries to transform before calling fit, we should handle it consistently within all transformers. So that would be a separate change.

Reply:
I agree that this should be covered in a generic way, probably in the `BaseTransformer` (BTW, the `HyperTransformer already covers it!).
In any case, I suggest changing the line to `if not self.categories_to_values` anyway, since in general `if not <iterable>` is the recommended pythonic shortcut for `if len(<iterable>) == 0` (I would even expect PyLint to fail on that one), and it is also slightly more efficient.

Reply:
I'm deleting this check. It seems to be causing too much confusion for something that should be deleted in the future either way.